internal/dag/cache.go

```diff
@@ -70,24 +72,40 @@ func (kc *KubernetesCache) init() {
 	kc.extensions = make(map[types.NamespacedName]*projectcontourv1alpha1.ExtensionService)
 }
 
+// Ingresses returns a set of valid Ingress objects that match proper
+// ingress class logic via annotations or IngressClasses.
+func (kc *KubernetesCache) Ingresses() []*v1beta1.Ingress {
+	ings := make([]*v1beta1.Ingress, 0, len(kc.ingresses))
+	for _, ing := range kc.ingresses {
+		if kc.matchesIngressClass(ing.Annotations, ing.Spec.IngressClassName) {
+			ings = append(ings, ing)
+		}
+	}
+	return ings
+}
+
 // matchesIngressClass returns true if the given Kubernetes object
 // belongs to the Ingress class that this cache is using.
-func (kc *KubernetesCache) matchesIngressClass(obj k8s.Object) bool {
+func (kc *KubernetesCache) matchesIngressClass(annotations map[string]string, ingressClassName *string) bool {
 
-	if !annotation.MatchesIngressClass(obj, kc.IngressClass) {
-		kind := k8s.KindOf(obj)
-		om := obj.GetObjectMeta()
+	// Check if the object has an annotation and that the annotation was found.
+	if found, classAnnotation := annotation.IngressClass(annotations); found {
+		return annotation.MatchesIngressClass(classAnnotation, kc.IngressClass)
+	}
 
-		kc.WithField("name", om.GetName()).
-			WithField("namespace", om.GetNamespace()).
-			WithField("kind", kind).
-			WithField("ingress.class", annotation.IngressClass(obj)).
-			Debug("ignoring object with unmatched ingress class")
+	if ingressClassName != nil {
+		// Check if this class matches a existing IngressClass.
+		_, ok := kc.ingressclasses[*ingressClassName]
+		return ok
+	}
```

Comment:
I thought I left a comment before, but I can't find it now :-/

IIUC, Contour will cache all IngressClass resources that match the controller name. Here, we will match any ingress class whose name matches the `Spec.IngressClassName`. So the result of this is that Contour will accept all resources which map to a Contour IngressClass regardless of the class name.

Did I get that right?